### DISCUSSION TITLE
Tree-sitter Grammar for Janet added to nvim-treesitter

### DISCUSSION THREAD

**Discussion by sogaiu** [janet-lang/janet#1118] [Show and tell]
The master branch of nvim-treesitter now has support for Janet.

It's possible that users of Conjure might notice some differences if this functionality is used.

Once the grammar is installed for nvim-treesitter, Conjure should automatically make use of it for form recognition / navigation [1].  If there were cases where forms were not-quite-recognized when not using tree-sitter, possibly this tree-sitter-based arrangement might make some difference [2].

Here's what it can look like (with TSPlayground open on the left):

![conjure-with-tree-sitter-janet-simple-new](https://user-images.githubusercontent.com/33044872/236173911-ab2ccd33-0361-48cd-a8f1-799d83489e99.png)

I've found using TSPlayground can be handy when trying to learn / decode programming languages I'm not so familiar with.

---

[1] I'm not a heavy user of Conjure or Neovim and haven't figured out differences in behavior (^^;

[2] If something works worse than before, please mention it -- though possibly Conjure's discord might be a better place for that.

**Comment by CosmicToast** [comment]
That's pretty nice!
One thing works worse than before (relative to the janet.vim plugin) though: keywords (`:a`) are harder to tell apart from "regular" keywords. At least with my setup.
Here's an example screenshot of `jurl/init.janet`:
<img width="752" alt="image" src="https://user-images.githubusercontent.com/5273047/236245804-f0d77142-096d-4ac8-b623-ca48074ad84f.png">
Outside of that though, it's much better and faster, so I'll be keeping it anyway, thanks! :)

**Reply by sogaiu** [reply]
Thanks for trying it out and the feedback.

I'm puzzled about keywords and (not built-in) symbols having the same highlighting (or at least looking the same).  That doesn't seem to happen in my setup (using gruvbox).

Below are links to the two lines that I think are responsible:

* [kwd_lit](https://github.com/nvim-treesitter/nvim-treesitter/blob/01c23a8036340af45cd90fa5b771ba16bd6eb152/queries/janet_simple/highlights.scm#L3)
* [sym_lit](https://github.com/nvim-treesitter/nvim-treesitter/blob/01c23a8036340af45cd90fa5b771ba16bd6eb152/queries/janet_simple/highlights.scm#L20)

One way to tweak things is to change the `@symbol`/  `@variable` value(s) in the linked lines to something else from [here](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights).  A bit hacky perhaps and the changes might get overwritten when updating nvim-treesitter.

IIUC, some level of customization is possible according to: https://github.com/nvim-treesitter/nvim-treesitter/tree/01c23a8036340af45cd90fa5b771ba16bd6eb152#adding-queries [1]

---

[1] That info seemed slightly incomplete -- I found [this](https://github.com/nvim-treesitter/nvim-treesitter/discussions/3729#discussioncomment-3990282) that mentions `;; extends` is necessary if using `after` queries.

**Reply by CosmicToast** [reply]
I took a somewhat closer look. I think your treesitter grammar is correct, but the issue lies more with nvim-treesitter.
If you look at `:h treesitter-highlight-groups`, you'll notice that they don't define a highlight group for `@symbol`, only `@variable` (as `Identifier`).

I added a custom `@symbol` highlight group to my theme, with great success.
I've opened [an issue](https://github.com/echasnovski/mini.nvim/issues/317) with `mini.base16` (the utility I use to generate my theme) to explicitly support `@symbol`.
Ideally, treesitter should export it as a built-in highlight group, but there doesn't seem to be one that matches (likely why it hasn't been done so far).

Thanks for giving all of the above context, it helped me figure out what was wrong much faster than I would have otherwise!

**Comment by CosmicToast** [comment]
I've been test driving it for a bit now (with the vimscript plugin disabled) and the main thing that quite bothers me is the lack of some important features.
For example, I can't seem to toggle comments (commentstring is empty) and indentation not working properly (even to the point of confusing parinfer).
It's still quite nice to have, but I'll be sticking to the vimscript plugin for now (but keeping this enabled!), using it only for highlighting.

**Reply by sogaiu** [reply]
Thanks for the detailed feedback.

As you metnioned there isn't much (if anything) apart from highlighting support.  I also find the indentation experience to be way better with [bakpakin's vimscript plugin](https://github.com/janet-lang/janet.vim) enabled [1].  Your setup of using the vimscript plugin for most features seems like the way to go.

---

[1] I don't know if anyone does this, but possibly an alternative is to periodically send buffer text to andrewchambers' [jfmt](https://github.com/andrewchambers/jfmt).

**Comment by sogaiu** [comment]
For future reference, I'm posting some relatively minimal configurations (`init.vim`) that seem to work for me.  Note that it may be necessary to modify these for a different plugin manager (or if no plugin manager).

Just using nvim-treesitter (as noted above, this only provides highlighting):

```vim
" Specify a directory for plugins
call plug#begin('~/.vim/plugged')
 
Plug 'nvim-treesitter/nvim-treesitter'
 
call plug#end()
 
" adapted:
"   https://github.com/janet-lang/janet.vim/blob/master/ftdetect/janet.vim
au BufRead,BufNewFile *.janet,*.jdn setlocal filetype=janet
 
lua <<EOF
require('nvim-treesitter.configs').setup {
  ensure_installed = "all",
  highlight = {
    enable = true,
    disable = { "css" },
  },
}
EOF
```

With janet.vim and nvim-treesitter (has various essentials including indentation):

```vim
" Specify a directory for plugins
call plug#begin('~/.vim/plugged')
 
Plug 'bakpakin/janet.vim'

Plug 'nvim-treesitter/nvim-treesitter'
 
call plug#end()
  
lua <<EOF
require('nvim-treesitter.configs').setup {
  ensure_installed = "all",
  highlight = {
    enable = true,
    disable = { "css" },
  },
}
EOF
```
